sql/core/src/main/scala/org/apache/spark/sql/execution/ExplainUtils.scala

```diff
@@ -0,0 +1,214 @@
+        i = i + 1
+        append(s"Subquery:$i Hosting operator id = " +
+          s"${getOpId(sub._1)} Hosting Expression = ${sub._2}\n")
+
+        // For each subquery expression in the parent plan, process its child plan to compute
+        // the explain output.
+        currentOperatorID = processPlanSkippingSubqueries(
+          sub._3,
+          append,
+          currentOperatorID)
+        append("\n")
+      }
+    } finally {
+      removeTags(plan)
+    }
+  }
+
+  /**
+   * Traverses the supplied input plan in a bottem-up fashion to produce the following two maps :
+   *    1. operator -> operator identifier
+   *    2. operator identifier -> operator
+   * Note :
+   *    1. Operator such as WholeStageCodegenExec and InputAdapter are skipped as they don't
+   *       appear in the explain output.
+   *    2. operator identifier starts at startIdx + 1
+   */
+  private def generateOperatorIDs(
+      plan: QueryPlan[_],
+      startOperatorID: Int,
+      operatorIDs: mutable.ArrayBuffer[(Int, QueryPlan[_])]): Int = {
+    var currentOperationID = startOperatorID
+    // Skip the subqueries as they are not printed as part of main query block.
+    if (plan.isInstanceOf[BaseSubqueryExec]) {
+      return currentOperationID
+    }
+    plan.foreachUp {
+      case p: WholeStageCodegenExec =>
+      case p: InputAdapter =>
+      case other: QueryPlan[_] =>
+        if (!other.getTagValue(QueryPlan.opidTag).isDefined) {
+          currentOperationID += 1
+          other.setTagValue(QueryPlan.opidTag, currentOperationID)
+          operatorIDs += ((currentOperationID, other))
+        }
+        other.innerChildren.foreach { plan =>
+          currentOperationID = generateOperatorIDs(plan,
+            currentOperationID,
+            operatorIDs)
+        }
+    }
+    currentOperationID
+  }
+  /**
+   * Traverses the supplied input plan in a top-down fashion to produce the following map:
+   *    1. operator -> whole stage codegen id
+   */
+  private def generateWholeStageCodegenIdMap(plan: QueryPlan[_]): Unit = {
+    // Skip the subqueries as they are not printed as part of main query block.
+    if (plan.isInstanceOf[BaseSubqueryExec]) {
+      return
+    }
+    var currentCodegenId = -1
+    plan.foreach {
+      case p: WholeStageCodegenExec => currentCodegenId = p.codegenStageId
+      case p: InputAdapter => currentCodegenId = -1
+      case other: QueryPlan[_] =>
+        if (currentCodegenId != -1) {
+          other.setTagValue(QueryPlan.codegenTag, currentCodegenId)
+        }
+        other.innerChildren.foreach { plan =>
+          generateWholeStageCodegenIdMap(plan)
+        }
+    }
+  }
+
+  /**
+   * Given a input plan, returns an array of tuples comprising of :
+   *  1. Hosting opeator id.
+   *  2. Hosting expression
+   *  3. Subquery plan
+   */
+  private def getSubqueries[T <: QueryPlan[T]](
+      plan: => QueryPlan[_],
+      subqueries: ArrayBuffer[(SparkPlan, Expression, SparkPlan)]): Unit = {
```

Comment:
instead of passing in a mutable map and return Unit, can we return a map directly?

Reply:
@cloud-fan Since this function is recursive, i think having it this way is more optimal ? Or else we may need to create more objects and do more copying ?

https://stackoverflow.com/questions/27886116/how-to-return-an-arraylist-with-an-recursive-function

Reply:
ah i see